clang/lib/Sema/SemaType.cpp

```diff
@@ -8479,10 +8480,131 @@ static void HandlePtrAuthQualifier(ASTContext &Ctx, QualType &T,
+    } else {
+      Expr::EvalResult Eval;
+      bool Result = AuthenticationOptionsArg->EvaluateAsRValue(Eval, Ctx);
+      if (Result && Eval.Val.isLValue()) {
+        auto *BaseExpr = Eval.Val.getLValueBase().dyn_cast<const Expr *>();
+        const StringLiteral *EvaluatedStringLiteral =
+            dyn_cast<StringLiteral>(const_cast<Expr *>(BaseExpr));
+        if (EvaluatedStringLiteral) {
+          CharUnits StartOffset = Eval.Val.getLValueOffset();
+          EvaluatedString = EvaluatedStringLiteral->getString().drop_front(
+              StartOffset.getQuantity());
+          OptionsString = EvaluatedString;
+          HasEvaluatedOptionsString = true;
+        }
+      }
+    }
+    if (!HasEvaluatedOptionsString) {
+      DiagnoseInvalidOptionsParameter(
+          "must be a string of comma separated flags");
+      return;
+    }
+    for (char Ch : OptionsString) {
+      if (Ch != '-' && Ch != ',' && !isWhitespace(Ch) && !isalpha(Ch)) {
+        DiagnoseInvalidOptionsParameter("contains invalid characters");
+        return;
+      }
+    }
+    HasEvaluatedOptionsString = true;
+    OptionsString = OptionsString.trim();
+    llvm::SmallVector<StringRef> Options;
+    if (!OptionsString.empty())
+      OptionsString.split(Options, ',');
+
+    for (unsigned Idx = 0; Idx < Options.size(); ++Idx) {
+      StringRef OptionString = Options[Idx].trim();
+      auto HandleOption = [&](auto Value, auto *Option,
+                              std::string *LastOption = nullptr) {
+        if (!*Option) {
+          *Option = Value;
+          if (LastOption)
+            *LastOption = OptionString;
+          return true;
+        }
+        bool IsAuthenticationMode =
+            std::is_same_v<decltype(Value), PointerAuthenticationMode>;
+        S.Diag(AuthenticationOptionsRange.getBegin(),
+               diag::err_ptrauth_repeated_authentication_option)
+            << !IsAuthenticationMode << OptionString
+            << (LastOption ? *LastOption : "");
+        return false;
+      };
 
-  if (IsInvalid) {
-    Attr.setInvalid();
-    return;
+      if (OptionString.empty()) {
+        bool IsLastOption = Idx == (Options.size() - 1);
+        DiagnoseInvalidOptionsParameter(
+            IsLastOption ? "has a trailing comma" : "contains an empty option");
+        continue;
+      }
+      std::optional<PointerAuthenticationOption> Option =
+          PointerAuthenticationOption::from(OptionString);
+      if (!Option) {
+        auto TrailingOption = OptionString.drop_until(isWhitespace).trim();
+        if (TrailingOption.empty()) {
+          S.Diag(AuthenticationOptionsRange.getBegin(),
+                 diag::err_ptrauth_unknown_authentication_option)
+              << OptionString;
+        } else
+          S.Diag(AuthenticationOptionsRange.getBegin(),
+                 diag::err_ptrauth_option_missing_comma)
+              << OptionString.take_until(isWhitespace);
+
+        IsInvalid = true;
+      }
```

Comment:
i think this whole loop would be a lot simpler if we looked first for end of the
current parameter (ie `find_if(OptionsString, [](char c) {return isWhitespace(c) || c == ','; }`) instead of splitting on `,` and then trying to decide if individual components are valid.